floors/fetcher.go

```diff
@@ -0,0 +1,294 @@
+		}
+		f.SetWithExpiry(configs.URL, floorData, cacheExpiry)
+	}
+
+	// Send to refetch channel
+	f.configReceiver <- FetchInfo{AccountFloorFetch: configs, FetchTime: time.Now().Add(time.Duration(configs.Period) * time.Second).Unix(), RefetchRequest: true}
+
+}
+
+// Stop terminates fetcher service can closes all the channels and destroys worker pool
+func (f *PriceFloorFetcher) Stop() {
+	close(f.done)
+	f.pool.Stop()
+	close(f.configReceiver)
+}
+
+func (f *PriceFloorFetcher) submit(fetchInfo *FetchInfo) {
+	status := f.pool.TrySubmit(func() {
+		f.worker(fetchInfo.AccountFloorFetch)
+	})
+	if !status {
+		heap.Push(&f.fetchQueue, fetchInfo)
+	}
+}
+
+func (f *PriceFloorFetcher) Fetcher() {
+	//Create Ticker of 5 minutes
+	ticker := time.NewTicker(time.Duration(refetchCheckInterval) * time.Second)
+
+	for {
+		select {
+		case fetchInfo := <-f.configReceiver:
+			if fetchInfo.RefetchRequest {
+				heap.Push(&f.fetchQueue, &fetchInfo)
+			} else {
+				if _, ok := f.fetchInProgress[fetchInfo.URL]; !ok {
+					f.fetchInProgress[fetchInfo.URL] = true
+					f.submit(&fetchInfo)
+				}
+			}
+		case <-ticker.C:
+			currentTime := time.Now().Unix()
+			for top := f.fetchQueue.Top(); top != nil && top.FetchTime <= currentTime; top = f.fetchQueue.Top() {
+				nextFetch := heap.Pop(&f.fetchQueue)
+				f.submit(nextFetch.(*FetchInfo))
+			}
+		case <-f.done:
+			glog.Info("Price Floor fetcher terminated")
+			return
+		}
+	}
+}
+
+func fetchAndValidate(config config.AccountFloorFetch) (*openrtb_ext.PriceFloorRules, int) {
+	floorResp, maxAge, err := fetchFloorRulesFromURL(config)
+	if err != nil {
+		glog.Errorf("Error while fetching floor data from URL: %s, reason : %s", config.URL, err.Error())
+		return nil, 0
+	}
+
+	if len(floorResp) > (config.MaxFileSize * 1024) {
+		glog.Errorf("Recieved invalid floor data from URL: %s, reason : floor file size is greater than MaxFileSize", config.URL)
+		return nil, 0
+	}
+
+	var priceFloors openrtb_ext.PriceFloorRules
+	if err = json.Unmarshal(floorResp, &priceFloors.Data); err != nil {
+		glog.Errorf("Recieved invalid price floor json from URL: %s", config.URL)
+		return nil, 0
+	} else {
+		err := validateRules(config, &priceFloors)
+		if err != nil {
+			glog.Errorf("Validation failed for floor JSON from URL: %s, reason: %s", config.URL, err.Error())
+			return nil, 0
+		}
+	}
+
+	return &priceFloors, maxAge
+}
+
+// fetchFloorRulesFromURL returns a price floor JSON and time for which this JSON is valid
+// from provided URL with timeout constraints
+func fetchFloorRulesFromURL(configs config.AccountFloorFetch) ([]byte, int, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(configs.Timeout)*time.Millisecond)
+	defer cancel()
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodGet, configs.URL, nil)
+	if err != nil {
+		return nil, 0, errors.New("error while forming http fetch request : " + err.Error())
+	}
+
+	httpResp, err := http.DefaultClient.Do(httpReq)
```

Comment:
Please do not use the default client. There is not much control available through this connection pool. I recommend you create a new floors fetcher specific pool and pass it down from `router.go`.

Reply:
Should we also declare transport? or it will be okay if we use default transport?